### DISCUSSION TITLE
waiting for lock file on build directory

### DISCUSSION THREAD

**Discussion by 9mm** [mrcjkb/rustaceanvim#444] [Q&A]
What's the best way to work on a rust project while also using this plugin?

Currently in my terminal I do:

```
 cargo watch -x 'run serve'
```

And then I let rustacean also work in neovim. I notice a lot of times I get this error though "waiting for lock file on build directory" as I'm assuming both are trying to compile at once.

What is the proper way to do this?

```lua
  {
    "mrcjkb/rustaceanvim",
    ft = { "rust" },
    init = function()
      vim.g.rustaceanvim = {
        tools = {
          test_executor = "background",
        },
        server = {
          default_settings = {
            ["rust-analyzer"] = {
              cargo = {
                allFeatures = true,
              },
              rustfmt = {
                extraArgs = { "+nightly" },
              },
            },
          },
          on_attach = function(client, bufnr)
            local lsp_map = function(mode, keys, func, desc)
              vim.keymap.set(mode, keys, func, { buffer = bufnr, desc = desc })
            end

            -- rust-lsp
            lsp_map("n", "K", function() vim.cmd.RustLsp({ "hover", "actions" }) end, "Rust hover docs")
            lsp_map("n", "J", function() vim.cmd.RustLsp("joinLines") end, "Rust join lines")
            lsp_map("n", "<Leader>ca", function() vim.cmd.RustLsp("codeAction") end, "Rust Code action")
            lsp_map("n", "<Leader>rue", function() vim.cmd.RustLsp("explainError") end, "Rust error explain")
            lsp_map("n", "<Leader>rud", function() vim.cmd.RustLsp("openDocs") end, "Rust docs")
            lsp_map("n", "<Leader>rum", function() vim.cmd.RustLsp("expandMacro") end, "Rust expand macro")
            lsp_map("n", "<Leader>rut", function() vim.cmd.RustLsp("testables") end, "Rust test")

            -- copy from lsp_config
            lsp_map("n", "gd", vim.lsp.buf.definition, "Goto definition")
            lsp_map("n", "gD", vim.lsp.buf.declaration, "Goto declaration")
            lsp_map("n", "gI", vim.lsp.buf.implementation, "Goto implementation")
            lsp_map("n", "go", vim.lsp.buf.type_definition, "Goto type definition")

            local format_sync_grp = vim.api.nvim_create_augroup("RustaceanFormat", {})
            vim.api.nvim_create_autocmd("BufWritePre", {
              buffer = bufnr,
              callback = function() vim.lsp.buf.format() end,
              group = format_sync_grp,
            })
          end,
        },
      }
    end,
  },
```

**Comment by 9mm** [comment]
Also... as a follow up question, is this what causes formatting to just stop working very often? things will format up to a point and then it just stops. I have to close the file and re-open it, then save it again, and formatting works again.

Ive read things about long lines etc, but i dont think thats it otherwise simply closing and opening the file again wouldnt magically fix it

**Comment by mrcjkb** [comment]
Hey 👋 

This is likely because rust-analyzer (which uses cargo) and cargo share the same target directory. 

You could try changing the target directory for one of the processes by setting the `CARGO_TARGET_DIR` environment variable.

Otherwise, you might be more lucky searching or asking in the [rust-analyzer discussions](https://github.com/rust-lang/rust-analyzer/discussions).

**Reply by 9mm** [reply]
Hmm ok. i did find this answer similarly: https://stackoverflow.com/a/72935539/794481

its this how i do that with rustacean?

```
        server = {
          default_settings = {
            ["rust-analyzer"] = {
              runnables = {
                extraArgs = {
                  "--target-dir",
                  "target/rls/",
                },
              },
              cargo = {
                allFeatures = true,
              },
              rustfmt = {
                extraArgs = { "+nightly" },
              },
            },
          },
```

Optionally, whats the best way to set env arg var (`CARGO_TARGET_DIR `) with rustaceon so my project doesnt use it, but rustacean does?

**Reply by mrcjkb** [reply]
That looks like it's the correct way to set it.

I'm not sure if Neovim passes its environment to the LSP server, but you could also try setting `vim.env.CARGO_TARGET_DIR`.
